bilge-impl/src/default_bits.rs

```diff
@@ -0,0 +1,88 @@
+use crate::shared::{self, unreachable, BitSize, fallback::Fallback};
+use proc_macro2::{Ident, TokenStream};
+use proc_macro_error::abort_call_site;
+use quote::quote;
+use syn::{Data, DeriveInput, Fields, Type};
+
+pub(crate) fn default_bits(item: TokenStream) -> TokenStream {
+    let derive_input = parse(item);
+    //TODO: does fallback need handling?
+    let (derive_data, _, name, ..) = analyze(&derive_input);
+
+    match derive_data {
+        Data::Struct(data) => generate_struct_default_impl(name, &data.fields),
+        Data::Enum(_) => abort_call_site!("use derive(Default) for enums"),
+        _ => unreachable(()),
+    }
+}
+
+fn generate_struct_default_impl(struct_name: &Ident, fields: &Fields) -> TokenStream {
+    let default_value = fields
+        .iter()
+        .map(|field| generate_default_inner(&field.ty))
+        .reduce(|acc, next| quote!(#acc | #next));
+
+    quote! {
+        impl ::core::default::Default for #struct_name {
+            fn default() -> Self {
+                let mut offset = 0;
+                let value = #default_value;
+                let value = <#struct_name as Bitsized>::ArbitraryInt::new(value);
+                Self { value }
+            }
+        }
+    }
+}
+
+fn generate_default_inner(ty: &Type) -> TokenStream {
+    use Type::*;
+    match ty {
+        // TODO?: we could optimize nested arrays here like in `struct_gen.rs`
+        Array(array) => {
+            let len_expr = &array.len;
+            let elem_ty = &*array.elem;
+            // generate the default value code for one array element
+            let value_shifted = generate_default_inner(elem_ty);
+            quote! {{
+                // constness: iter, array::from_fn, for-loop, range are not const, so we're using while loops
+                let mut acc = 0;
+                let mut i = 0;
+                while i < #len_expr {
+                    // for every element, shift its value into its place
+                    let value_shifted = #value_shifted;
+                    // and bit-or them together
+                    acc |= value_shifted;
+                    i += 1;
+                }
+                acc
+            }}
+        },
+        Path(path) => {
+            let field_size = shared::generate_type_bitsize(ty);
+            // u2::from(HaveFun::default()).value() as u32;
+            quote! {{
+                let as_int = <#path as Bitsized>::ArbitraryInt::from(#path::default()).value();
```

Comment:
if `#path` does not implement `Default`, we get "no variant or associated item named `default` found for...". and unfortunately the span for this error is the `#[bitsize(...)]` attribute.

to make this message match the one generated by `std`'s `#[derive(Default)]`, we can add `assume_default<#path>()` here. unfortunately if this was, say, a tuple, this would generate a message for each element (with duplicates). AND the original "no variant or associated item" is still generated.

might be possible to generate the assert for each field of the original struct instead. I don't know if any of this even matters.

Reply:
says "the following trait defines an item `default`, perhaps you need to implement it" on my end, but we can fix this even more:
`<#path as ::core::default::Default>::default()`
-> "the trait bound `HaveFun: std::default::Default` is not satisfied
the trait `std::default::Default` is not implemented for `HaveFun`"

Reply:
oh, beautiful.